### ISSUE TITLE
Change of binder behavior whan narrowing to unrelated type

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#18967]
The behavior in this example was changed in #18538:
```py
def any():
    return 1

def f() -> None:
    x = "a"
    x = any()
    assert isinstance(x, int)
    reveal_type(x)  # "Never", but used to be unreachable (no message)
```
In the past, the code after `isinstance` was silently unchecked. Now the inferred type is `Never`, which will likely generate false positives. Neither behavior is great, but users may consider the switch from false negatives to false positives to be a regression. I'm not sure what we should do here. If we'd allow variables to be redefined freely, there would be no issue.

cc @ilevkivskyi who authored the PR that changed the behavior

**Comment by A5rocks** [comment]
In https://github.com/python/mypy/pull/18707 I changed things such that narrowing to `Never` (or assigning it in any way, really) would mark a block as unreachable. That seemed to work decently well.

I would be fine to extract that if it's a good idea.

**Comment by ilevkivskyi** [comment]
Marking this code unreachable is a possible solution, but it may cause undesired behavior in other cases. Another possible option would be to replace `Never` with known restriction (i.e. `int` in this example), it will likely have a more limited impact. I can try this now, and see if there any adverse effects.

**Comment by ilevkivskyi** [comment]
I will double-check (superficially good) fallout in https://github.com/python/mypy/pull/18972 later today or tomorrow.
